x/wasm/internal/keeper/keeper_test.go

```diff
@@ -451,6 +453,116 @@ func TestExecuteWithStorageLoop(t *testing.T) {
+	require.NoError(t, err)
+
+	_, _, anyAddr := keyPubAddr()
+	initMsg := InitMsg{
+		Verifier:    fred,
+		Beneficiary: anyAddr,
+	}
+	initMsgBz, err := json.Marshal(initMsg)
+	require.NoError(t, err)
+	specs := map[string]struct {
+		admin                sdk.AccAddress
+		overrideContractAddr sdk.AccAddress
+		caller               sdk.AccAddress
+		codeID               uint64
+		migrateMsg           []byte
+		expErr               *sdkerrors.Error
+	}{
+		"all good with same code id": {
+			admin:  creator,
+			caller: creator,
+			codeID: originalContractID,
+		},
+		"all good with new code id": {
+			admin:  creator,
+			caller: creator,
+			codeID: newContractID,
+		},
+		"prevent migration when admin was not set on instantiate": {
+			caller: creator,
+			codeID: originalContractID,
+			expErr: sdkerrors.ErrUnauthorized,
+		},
+		"prevent migration when not admin": {
+			caller: creator,
+			admin:  fred,
+			codeID: originalContractID,
+			expErr: sdkerrors.ErrUnauthorized,
+		},
+		"fail with non existing code id": {
+			admin:  creator,
+			caller: creator,
+			codeID: 99999,
+			expErr: sdkerrors.ErrInvalidRequest,
+		},
+		"fail with non existing contract addr": {
+			admin:                creator,
+			caller:               creator,
+			overrideContractAddr: anyAddr,
+			codeID:               originalContractID,
+			expErr:               sdkerrors.ErrInvalidRequest,
+		},
+		"fail when migration caused error": {
+			admin:                creator,
+			caller:               creator,
+			overrideContractAddr: anyAddr,
+			codeID:               originalContractID,
+			migrateMsg:           bytes.Repeat([]byte{0x1}, 7), // condition hard coded in stub: >6 = error
+			expErr:               sdkerrors.ErrInvalidRequest,
+		},
+	}
+	var (
+		builtIntoGoCosmWasmStubGas  = sdk.Gas(10000)
+		builtIntoGoCosmWasmStubData = []byte(("my-migration-response-data"))
+	)
+	for msg, spec := range specs {
+		t.Run(msg, func(t *testing.T) {
+			ctx = ctx.WithBlockHeight(ctx.BlockHeight() + 1)
+			addr, err := keeper.Instantiate(ctx, originalContractID, creator, spec.admin, initMsgBz, "demo contract", nil)
+			require.NoError(t, err)
+			if spec.overrideContractAddr != nil {
+				addr = spec.overrideContractAddr
+			}
+			gasBefore := ctx.GasMeter().GasConsumed()
+			res, err := keeper.Migrate(ctx, addr, spec.caller, spec.codeID, spec.migrateMsg)
+			require.True(t, spec.expErr.Is(err), "expected %v but got %+v", spec.expErr, err)
+			if spec.expErr != nil {
+				return
+			}
+			gasAfter := ctx.GasMeter().GasConsumed()
+			assert.Greater(t, gasAfter-gasBefore, builtIntoGoCosmWasmStubGas)
+			assert.Equal(t, builtIntoGoCosmWasmStubData, res.Data)
+			cInfo := keeper.GetContractInfo(ctx, addr)
+			assert.Equal(t, spec.codeID, cInfo.CodeID)
+			assert.Equal(t, originalContractID, cInfo.PreviousCodeID)
+			assert.Equal(t, types.NewCreatedAt(ctx), cInfo.LastUpdated)
+			// TODO: check contract store was updated by migration code (impl also in contract)
```

Comment:
Good TODOs that can probably wait until a real migration implementation.

Reply:
Ja, either I have a mock in wasmd or I need to wait for a test impl in go-cosmwasmd and a demo contract. A mock would be easier for me to do but with a new PR if 👍